includes/class-wc-admin-notes-order-milestones.php

```diff
@@ -0,0 +1,251 @@
+ *
+ * Adds a note to the merchant's inbox when certain order milestones are reached.
+ *
+ * @package WooCommerce Admin
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * WC_Admin_Notes_Order_Milestones
+ */
+class WC_Admin_Notes_Order_Milestones {
+	/**
+	 * Name of the "first order" note.
+	 */
+	const FIRST_ORDER_NOTE_NAME = 'wc-admin-first-order';
+
+	/**
+	 * Name of the "ten orders" note.
+	 */
+	const TEN_ORDERS_NOTE_NAME = 'wc-admin-ten-orders';
+
+	/**
+	 * Name of the "other milestones" note.
+	 */
+	const ORDERS_MILESTONE_NOTE_NAME = 'wc-admin-orders-milestone';
+
+	/**
+	 * Option key name to store last order milestone.
+	 */
+	const LAST_ORDER_MILESTONE_OPTION_KEY = 'wc_admin_last_orders_milestone';
+
+	/**
+	 * Hook to process order milestones.
+	 */
+	const PROCESS_ORDERS_MILESTONE_HOOK = 'wc_admin_process_orders_milestone';
+
+	/**
+	 * Allowed order statuses for calculating milestones.
+	 *
+	 * @var array
+	 */
+	protected $allowed_statuses = array(
+		'pending',
+		'processing',
+		'completed',
+	);
+
+	/**
+	 * Orders count cache.
+	 *
+	 * @var int
+	 */
+	protected $orders_count = null;
+
+	/**
+	 * Further order milestone thresholds.
+	 *
+	 * @var array
+	 */
+	protected $milestones = array(
+		100,
+		250,
+		500,
+		1000,
+		5000,
+		10000,
+		500000,
+		1000000,
+	);
+
+	/**
+	 * Delay hook attachment until after the WC post types have been registered.
+	 *
+	 * This is required for retrieving the order count.
+	 */
+	public function __construct() {
+		add_action( 'woocommerce_after_register_post_type', array( $this, 'init' ) );
+	}
+
+	/**
+	 * Hook everything up.
+	 */
+	public function init() {
+		if ( ! wp_next_scheduled( self::PROCESS_ORDERS_MILESTONE_HOOK ) ) {
+			wp_schedule_event( time(), 'hourly', self::PROCESS_ORDERS_MILESTONE_HOOK );
+		}
+
+		add_action( 'wc_admin_installed', array( $this, 'backfill_last_milestone' ) );
+
+		if ( 10 <= $this->get_orders_count() ) {
+			add_action( 'woocommerce_new_order', array( $this, 'first_two_milestones' ) );
+		}
+
+		add_action( self::PROCESS_ORDERS_MILESTONE_HOOK, array( $this, 'other_milestones' ) );
+	}
+
+	/**
+	 * Get the total count of orders (in the allowed statuses).
+	 *
+	 * @return int Total orders count.
+	 */
+	public function get_orders_count() {
+		if ( is_null( $this->orders_count ) ) {
+			$status_counts       = array_map( 'wc_orders_count', $this->allowed_statuses );
+			$this->orders_count  = array_sum( $status_counts );
+		}
+
+		return $this->orders_count;
+	}
+
+	/**
+	 * Add a milestone notes for the store's first order and first 10 orders.
+	 *
+	 * @param int $order_id WC_Order ID.
+	 */
+	public function first_two_milestones( $order_id ) {
+		$order = wc_get_order( $order_id );
+
+		// Make sure this is the first pending/processing/completed order.
+		if ( ! in_array( $order->get_status(), $this->allowed_statuses ) ) {
+			return;
+		}
+
+		$orders_count = $this->get_orders_count();
+
+		if ( 1 === $orders_count ) {
```

Comment:
I'm not seeing where `set_last_milestone()` is called for these two instances of first and tenth order?

Reply:
So the only reason I kind of let my brain mentally wander down that path was a scenario where I get my first order... and what if that is me testing out my site ( which we went around in circles on so don't want to go there too much again ) and then I get the notif, then delete that order, and a new order comes in. Edge case I suppose, but wouldn't that trigger the notif again for the first order? Same for a 10th order getting placed and one being deleted.

Reply:
Yes in your example the notification would get triggered again if you deleted your test order. But.. wouldn't that be a good thing?